book/python/using-connectors.md

```diff
@@ -0,0 +1,193 @@
+
+Here is an example of what this looks like:
+
+```python
+def application_setup(args):
+    ab = wallaroo.ApplicationBuilder("Celsius to Fahrenheit")
+
+    ab.source_connector("celsius_feed",
+        encoder=encode_feed,
+        decoder=decode_feed)
+
+    ab.sink_connector("fahrenheit_conversion",
+        encoder=encode_conversion,
+        decoder=decode_conversion)
+
+    ab.new_pipeline("Celsius Conversion", "celsius_feed")
+    ab.to(multiply)
+    ab.to(add)
+    ab.to_sink("fahrenheit_conversion")
+    return ab.build()
+```
+
+You can follow along in this example by going to [github](https://github.com/Wallaroo/wallaroo/tree/{{book.wallaroo_version}}/examples/python/celsius_connectors/).
+
+We have introduced two new application builder methods for declaring connectors: `source_connector` and `sink_connector`. These allow you to describe both ends of the connection so the connector script can encode or decode data in a way that's compatible with your application. Keeping these in one place helps ensure that it's easy to keep these in sync.
+
+In our example we can look at the celsius feed's encoder and decoder functions:
+
+```python
+@wallaroo.experimental.stream_message_encoder
+def encode_feed(data):
+    return data
+
+@wallaroo.experimental.stream_message_decoder
+def decode_feed(data):
+    return struct.unpack(">f", data)[0]
+```
+
+Here we've defined two functions and use `wallaroo.experimental.stream_message_encoder` and `wallaroo.experimental.stream_message_decoder` as decorators. These decorators will handle framing and delivery of your data, so all you need to ensure is that the bytes represent the single message you want to deliver.
+
+A good observer will note that we're just returning the data in this example's `encode_feed` function. In this case we're receiving data that's already a packed floating point number so there is no need to reencode it. This allows you to pass along data as-is when it's already in the correct format.
+
+Returning to the application setup code, we should look at the pipeline defintion to understand how all of this comes together. First is the `new_pipeline` call:
+
+```python
+ab.new_pipeline("Celsius Conversion", "celsius_feed")
+```
+
+Instead of the traditional `TCPSourceConfig` instance that we might have used before for a basic TCP stream, we now use the name of the connector. This must match the type of connector; in this case, new_pipeline expects a source, so we should declare `"celsius_feed"` using `source_connector`.
+
+Likewise, our sink has adopted a named connector:
+
+```python
+ab.to_sink("fahrenheit_conversion")
+```
+
+This matches up with our call to `sink_connector` and it has its own encoder and decoder functions as well.
+
+Notice how we don't couple the specifics of the medium used by the connector and instead focus on how we interpret incoming messages in terms of their serialization. This allows applications to evolve and mix their infrastucture over time.
```

Comment:
This is an interesting point. I feel it could be elaborated on. (Not now, just noting this is an interesting post that could form the basis of other documentation, rationale or a blog post)

Reply:
I agree. I had more here in a past draft but cut it as it became a distraction from the rest of the flow.